src/Onnx/OnnxToFeld.hs

```diff
@@ -236,42 +264,142 @@ mkInitReadFile hf initGroups = L.concat $ start : map initVec initGroups ++ map
+allocReadTensor :: L.ByteString -> L.ByteString -> Int -> TD.DataType -> L.ByteString
+allocReadTensor file prefix nDims elemTT = L.unlines code <> "\n"
+  where code = map ("  " <>) $ readDims ++ [alloc] ++ readElems
         elemTC = showCElemT elemTT
-        elemTT = tiType ti
-        dims = D.toList $ tiDims ti
-        initSizes []  = error "Zero dimensional tensor not implemented"
-        initSizes [d] = [awlPrefix <> ".length = " <> show d <> ";"]
-        initSizes ds  = [ prefix <> ".member2.member" <> show j <> " = " <> show n <> ";"
-                        | (n,j) <- zip ds [1 :: Int ..]
-                        ] ++ [awlPrefix <> ".length = " <> show (product dims) <> ";"]
-        readDims = "fscanf(wfile, \" " <> elemTH <> " " <> show (length dims) <> " " <> L.unwords (map show dims) <> "\");"
-        alloc awl = awl <> ".buffer = malloc(" <> awl <> ".length * sizeof(" <> elemTC <> "));"
-        readElems awl = [ "for(int i = 0; i < " <> awl <> ".length; i++) {"
-                        , "  fscanf(wfile, " <> scanFormat elemTT <> ", " <> awl <> ".buffer + i);"
-                        , "}"
-                        ]
+        (awlPrefix, readDims) = initTensorDims file prefix nDims
+        alloc = awlPrefix <> ".buffer = malloc(" <> awlPrefix <> ".length * sizeof(" <> elemTC <> "));"
+        readElems = [ "for(int i = 0; i < " <> awlPrefix <> ".length; i++) {"
+                    , "  check(fscanf(" <> file <> ", " <> scanFormat elemTT <> ", " <> awlPrefix <> ".buffer + i), 1);"
+                    , "}"
+                    ]
+
+-- | Initialize the size of tensor from the dimensions in a file
+initTensorDims :: L.ByteString -> L.ByteString -> Int -> (L.ByteString, [L.ByteString])
+initTensorDims file prefix nDims = go nDims
+  where go 1 = (prefix, [mkScanf file [prefix <> ".length"]])
+        go n = (prefix <> ".member1",
+                [mkScanf file $ reverse vars, -- The Syntactic instance for Pull stores extents in reverse order
+                 prefix <> ".member1.length = " <> L.intercalate " * " vars <> ";"])
+           where vars = [prefix <> ".member2.member" <> show i | i <- [1 :: Int .. n]]
+
+-- | Generate a call to fscanf reading the tensor dimensions
+mkScanf :: L.ByteString -> [L.ByteString] -> L.ByteString
+mkScanf file vars = "check(fscanf(" <> file <> ", " <> format <> ptrs <> "), " <> show n <> ");"
+  where format = "\" " <> show n <> L.concat (replicate n " %u") <> "\""
+        ptrs = L.concat [", &" <> v | v <- vars]
+        n = length vars
+
+-- | Generate a variable name, a type and code for printing output
+mkOutput :: [V.ValueInfoProto] -> (L.ByteString, L.ByteString, L.ByteString)
+mkOutput outputs = go outputs
+  where go [vip] = (v, fromString $ renderType t, printTensor "stdout" v dt)
+          where dt = vipType vip
+                t = uncurry tensorToType dt
+                v = mangle $ vipName vip
+        go _ = error "Only a single output currently supported"
```

Comment:
Prefix the error message with `functionName: ` to simplify finding which function that called error.

Reply:
Done.